tests/profiler/test_rule_state.py

```diff
@@ -20,27 +21,33 @@ def semantic_rule_state():
     return RuleState(
         active_domain={"domain_kwargs": {"column": "Age"}, "semantic_type": "numeric"},
         domains=[{"domain_kwargs": {"column": "Age"}, "semantic_type": "numeric"}],
-        parameters={"f45a40fda1738351c5e67a0aa89c2c7c": {"mean": 5.0}},
-        variables={"false_positive_threshold": 0.01},
+        parameters={
+            "f45a40fda1738351c5e67a0aa89c2c7c": Parameter(
```

Comment:
This `f45...` is what I was describing above. Is this the `parameter_id` or is `mean` the `parameter_id`?  Or is this the `domain_id` (if that is the case, I think this solves some of our problems, though I would then expect that there should be multiple parameters allowed per domain, and so this should be a list or dictionary).

Reply:
@talagluck In my understanding, this `f45...` string is the `domain ID`.  Thus far, the `domain` is represented by a dictionary.  Maybe it makes sense to have a typed object for it?  Let us discuss this first, before jumping into implementation.  (We have the `DomainBuilder` object.)  Our `RuleState` object has the notions of all `domains` (as a list) and the `active_domain`; for the `active_domain`, `RuleState` provides the `active_domain_id` property (using `IDDict` to generate it, guaranteed to be unique).  The `parameters` property in `RuleState` is a dictionary, whose key is the domain ID with the `Parameter` object being the value.  The test is constructed according to this structure.

Reply:
Writing this out to clarify:

- We currently have one domain builder per rule (which can encompass many domains, like all numeric columns, or all columns ending in `_id`.
- RuleState encompasses state for all rules.
- What does the active_domain in RuleState do?
- If I have a dataset with 3 columns - `a`, `b`, and `c`- all of which are numeric. I create a rule that says for each numeric column, I would like to get the min and the max value. We now have three mins, and three maxes, for columns `a`, `b`, and `c`. Can these all be stored in the above RuleState?

Reply:
@talagluck -- On bullet 3: the `active_domain` plays a role in returning the value of the parameter -- when the parameter name fits a certain reserved keyword, such as `$domain.domain_kwargs`.  Also, if the parameter name does not refer a variable (for variables, the parameter name must start with `$variables.`), then the starting (at the top level of the parameter naming hierarchy -- i.e., the entire fully-qualified parameter name string) is determined by using the ID of the active domain as the key to the `parameters` dictionary (domain_id to Parameter object dict), passed to the constructor.  All this suggests to me that the term `Parameter` may not be the best fit for the semantics of this object.  Happy to discuss.

On bullet 2: I could be wrong, but it seems that the answer is affirmative if we consider that the starting parameter name is a fully-qualified parameter name, meaning the dot punctuated set of syllables.  So it results in a hierarchy of keywords and the `Parameter` references, the tail one being the very metric, `min` or `max`, but the entire object trees are distinct.  That is assuming that prior to using `RuleState`, we have created the `parameters` correctly, which we would pass to `RuleState` upon instantiation.  Would be happy to discuss to work through details.